### PR TITLE
fix(libutil): normalize path components in `CanonPath::operator/`

### PR DESCRIPTION
**Pull request by lovesegfault**
## Motivation

Previously, `CanonPath::operator/(string_view)` would call `push()`
directly, which asserts that components don't contain `.` or `..`. This
caused cryptic assertion failures when code constructed paths like:

```
CanonPath("/foo") / "."
```

The assertion message was unhelpful:

```
nix: ../canon-path.cc:82: void nix::CanonPath::push(std::string_view): Assertion `c != "." && c != ".."' failed.
```

This commonly occurred when using Nix path literals with trailing `/.`:

```nix
builtins.getFlake (toString ~/src/nix/.)
```

This is because `operator/` was bypassing the path normalization that
the `CanonPath` constructor provides.

With this change, we use `CanonPath(c, *this)` constructor which
properly normalizes the component relative to the current path, handling
`.`, `..`, and multiple slashes.

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libutil-tests/canon-path.cc [hunk 1]
```diff
@@ -132,6 +132,57 @@ TEST(CanonPath, concat)
...
+    {
+        CanonPath p("/foo/bar");
+        ASSERT_EQ((p / "../baz").abs(), "/foo/baz");
+    }
+
+    // Test edge case: trailing "/."
+    {
+        CanonPath p("/foo");
+        ASSERT_EQ((p / "bar/.").abs(), "/foo/bar");
+    }
```
Comment (xokdvium) [inline src/libutil-tests/canon-path.cc]: Can you rewrite this with `INSTANTIATE_TEST_SUITE_P`? That would be much more compact

File: src/libutil/canon-path.cc [hunk 2]
```diff
@@ -87,9 +87,9 @@ void CanonPath::push(std::string_view c)
 
 CanonPath CanonPath::operator/(std::string_view c) const
 {
-    auto res = *this;
-    res.push(c);
-    return res;
+    // Normalize the component by constructing a CanonPath from it
+    // relative to this path, which will handle '.', '..', and multiple slashes properly
+    return CanonPath(c, *this);
```
Comment (xokdvium) [inline src/libutil/canon-path.cc]: I imagine this would be much less efficient

File: src/libutil-tests/canon-path.cc (hunk 1 above)
Comment (lovesegfault) [reply src/libutil-tests/canon-path.cc]: done! i keep forgetting to use that, my bad :)

File: src/libutil/canon-path.cc (hunk 2 above)
Comment (lovesegfault) [reply src/libutil/canon-path.cc]: i tried doing it only when necessary, take a look and let me know what you think?

File: src/libutil/canon-path.cc [hunk 3]
```diff
@@ -87,9 +87,16 @@ void CanonPath::push(std::string_view c)
...
 {
-    auto res = *this;
-    res.push(c);
-    return res;
+    // Fast path: if the component is simple (no slashes, not "." or ".."),
+    // we can use push() directly without normalization
+    if (c.find('/') == c.npos && c != "." && c != "..") {
+        auto res = *this;
+        res.push(c);
+        return res;
```
Comment (xokdvium) [inline src/libutil/canon-path.cc]: Not sure about this. The rhs of the operator is explicitly stated to not accept the segments with `/`. Also this might be the downstream that needs fixing.

**Review by xokdvium — CHANGES_REQUESTED** [review]
Let's put this up for triage

File: src/libutil/canon-path.cc (hunk 2 above)
Comment (lovesegfault) [reply src/libutil/canon-path.cc]: (we talked on matrix, see new approach)

**Comment by lovesegfault** [comment]
the question for triage

> Also maybe it's the downstream code that needs fixing. I think the idea for CanonPath (and this function) was to not accept . .. at all

**Comment by lovesegfault** [comment]
I can no longer repro this, so i'm closing it. If someone hits this again, this PR does seem to be the right fix.
